src/segment_reader.rs

```diff
@@ -342,6 +346,157 @@ impl AsyncSegmentReaderImpl {
     }
 }
 
+/// A wrapper around the AsyncSegmentReader.
+///
+/// It maintains a buffer and prefetches data to fill that buffer in the background so that
+/// every read call can read from the local buffer instead of waiting for server response.
+pub(crate) struct AsyncSegmentReaderWrapper {
+    buffer: Vec<u8>,
```

Comment:
Random thought: What if instead of having a vec of bytes that we copy replies out of, we just held a vec of replies?

Reply:
That's a very good idea. I have updated the implementation.